### PR TITLE
[BB-23] Add possibility to run compatibility check

### PR DESCRIPTION
**Pull request by mzanellazextras**

### CODE REVIEW

**Review by zextrasGiovanniDefacci — APPROVED** [review]

File: buildzaltodistdir [hunk 1]
```diff
@@ -0,0 +1,11 @@
...
+
+set -e
+
+N=4
+i=0
+for j in $@; do
+  ((i=$i+1)); (($i%$N==0)) && wait
+   mvn install -Pcompatibility-check-output-dir -Dzimbra.version=$j &
+done
+wait
```
[the remark this answers is not in the document]
Comment (Polpetta) [reply buildzaltodistdir:18]: Ok, it is waiting for the mvn command to exit. But this doesn't work well when you press Ctrl+C for exiting the process. On top of that you would to save the PID of every execution and call wait on all of them, but it is still not elegant.
You need to add a `trap` if you want this to behave as one expects, or you need a proper program to do that, e.g. `parallel`

File: buildzaltodistdir [hunk 2]
```diff
@@ -0,0 +1,18 @@
...
+  fi
+fi
+
+N=4
+i=0
+for j in ${params[@]}; do
+  ((i=$i+1)); (($i%$N==0)) && wait
+   mvn install -Pcompatibility-check-output-dir -Dzimbra.version=$j &
+done
+wait
```
Comment (Polpetta) [inline buildzaltodistdir:18]: I've passed `shellcheck` and used `parallel` to run a parallel number of mvn invocations equal to half the number of processors one has, and removed the workarounds with `wait`. Now shellchecks doens't complain anymore and the script works like a charm, Ctrl+C handling included :fire: 
```suggestion
params=( "$@" )
concurrent_jobs="$(($(nproc)/2))"

echo "Performing ${concurrent_jobs} jobs at the same time"

if [ "${#}" == 1 ]; then
    if [ "${1}" == "all" ]; then
        params=(8.6.0 8.7.0 8.7.1 8.7.2 8.7.3 8.7.4 8.7.5 8.7.6 8.7.7 8.7.9 8.7.10 8.7.11 8.8.2 8.8.3 8.8.4 8.8.5 8.8.6 8.8.7 8.8.8 8.8.9 8.8.10 8.8.11 8.8.12 8.8.15 9.0.0)
    fi
fi

printf '%s\0' "${params[@]}" | parallel -kj"${concurrent_jobs}" -0 "mvn install -Pcompatibility-check-output-dir -Dzimbra.version={} || echo 'error while processing Openzal for Zimbra {}'; exit 1"

```

The output is printed according the input provided, i.e if the input is `8.6.0 8.7.7 8.8.15` the resulting output would be respectively for `8.6.0` and so on, while still executing in parallel. To grab an error you just have to execute the script and do `| grep 'error while processing'` :wink:

File: buildzaltodistdir [hunk 3]
```diff
@@ -0,0 +1,11 @@
+#!/bin/bash
+
+set -e
+
+N=4
+i=0
+for j in $@; do
+  ((i=$i+1)); (($i%$N==0)) && wait
```
[the remark this answers is not in the document]
Comment (Polpetta) [reply buildzaltodistdir:15]: Please ignore this suggestion, it is outdated

File: buildzaltodistdir [hunk 4]
```diff
@@ -0,0 +1,11 @@
...
+   mvn install -Pcompatibility-check-output-dir -Dzimbra.version=$j &
```
[the remark this answers is not in the document]
Comment (Polpetta) [reply buildzaltodistdir:16]: Ignore this suggestion, is it outdated

**Review by Polpetta — CHANGES_REQUESTED** [review]
Please fix `buildzaltodistdir` accordingly. Everything else is ok :+1:

File: buildzaltodistdir (hunk 1 above)
[the remark this answers is not in the document]
Comment (federicorispo) [reply buildzaltodistdir:18]: I would remove wait as well

**Review by federicorispo — CHANGES_REQUESTED** [review]

**Comment by federicorispo** [comment]
Idea: what if we moved the `japi-compliance-checker` tool into a separate project containing the bash script to launch it by hand? That way we wouldn't be forced to have a bash script that launches a jar that launches a perl script!

For what I know this check is done manually sometimes

**Pull request closed without merging**